feat: new functions in stdlib from `stdlib.fc` and `math.fc`

Gusarich opened the pull request:
<!--
IMPORTANT:
If your PR doesn't close a particular issue, please, create the issue first and describe the whole context: what you're adding/changing and why you're doing so. And only then open the Pull Request, which would close that issue!
-->

## Issue

Closes #770
Towards #1029 (only `myCode()`)

## Checklist

- [x] I have updated CHANGELOG.md
- [x] I have documented my contribution in `docs/` and made the build locally
- [x] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases
- [x] I have run all the tests locally and no test failure was reported
- [x] I have run the linter, formatter and spellchecker
- [x] I did not do unrelated and/or undiscussed refactorings

## Current State

- [x] All stdlib.fc functions are present in Tact stdlib in some form, with some exceptions:
  - no tuple manipulation functions, because there is no tuple support in Tact. Do we really need them? @anton-trunov 
  - no `minmax` and no `slice_bits_refs` because we don't have tensor types in Tact and it would be useless: #1116
  - no `set_data` and `get_data` because Tact manages persistent storage by itself
  - no `bless` and other continuation functions because Tact manages control flow itself and for running some arbitrary code it's better to use RUNVM which doesn't require continuations parameters
  - no `raw_reserve_extra` because it requires an `uint32 -> varuint32` dictionary and we only support `varuint16` for values currently: #1117
- [x] All mathlib.fc functions are present in Tact stdlib in some form, with some exceptions:
  - no fixed-point arithmetic related functions: #1118
  - no `nan` and `is_nan` because we don't support NaN in Tact. Do we really need them? @anton-trunov 
  - no `sub_rev` because it is useless in our case
  - no `...MOD` instructions because they return two values as a tensor and it won't be very developer-friendly in Tact to make them return tuples in my opinion: #1116
- [x] All new functions are covered with tests

Gusarich committed:
feat: add missing functions from `stdlib.fc`

Gusarich committed:
feat: add `sgn` math function

Gusarich committed:
chore: update changelog

novusnota commented on stdlib/stdlib.fc:
```diff
@@ -645,3 +645,22 @@ builder store_builder(builder to, builder from) asm "STBR";
+int get_original_fwd_fee(int workchain, int fwd_fee) asm(fwd_fee workchain) "GETORIGINALFWDFEE";
+int my_storage_due() asm "DUEPAYMENT";
+
+tuple get_fee_cofigs() asm "UNPACKEDCONFIGTUPLE";
```
```suggestion
tuple get_fee_configs() asm "UNPACKEDCONFIGTUPLE";
```

novusnota commented on stdlib/stdlib.fc:
```diff
@@ -440,7 +440,7 @@ builder store_ref(builder b, cell c) asm(c b) "STREF";
 builder store_slice(builder b, slice s) asm "STSLICER";
 
-;;; Stores (serializes) an integer [x] in the range `0..2^120 − 1` into `builder` [b].
+;;; Stores (serializes) an integer [x] in the range `0..2^128 − 1` into `builder` [b].
```
These things keep bubbling up :)

Gusarich replied:
oh, again, haha 😆
it's from the copy-pasting the latest stdlib from Core team projects

Gusarich replied:
fixed

novusnota commented on stdlib/stdlib.fc:
```diff
@@ -645,3 +645,22 @@ builder store_builder(builder to, builder from) asm "STBR";
 
 ;;; Retrieves code of smart-contract from c7
 cell my_code() asm "MYCODE";
+
```
Maybe, it would be handy to visually separate the regular `stdlib.fc` (from TON monorepo) and the additions you make specifically for Tact's use. Like, `;;; --- Tact-adjusted ---`, or something

Gusarich replied:
well, there's nothing adjusted specifically for Tact. I took the "newer" stdlib from one of the recent core projects and removed the stuff that they put there for their FunC code (helper functions for building cells and some constants already existing in Tact etc)

novusnota commented on stdlib/std/math.tact:
```diff
@@ -59,4 +59,6 @@ native log(num: Int, base: Int): Int;
\ No newline at end of file
+asm fun pow2(exp: Int): Int { POW2 }
+
+asm fun sgn(x: Int): Int { SGN }
```
Maybe, name it `sign()` for it to look less cryptic and be simpler to type

```suggestion
asm fun sign(x: Int): Int { SGN }
```

novusnota commented on stdlib/std/cells.tact:
```diff
@@ -147,4 +157,19 @@ inline fun emptyCell(): Cell {
+
+asm fun computeSliceDataSize(slice: Slice, maxCells: Int): DataSize { SDATASIZE TRIPLE }
+
+asm fun cellDepth(cell: Cell): Int { CDEPTH }
```
This should probably be an extension function

```suggestion
asm extends fun depth(self: Cell): Int { CDEPTH }
```

novusnota commented on stdlib/std/cells.tact:
```diff
@@ -105,6 +107,14 @@ asm extends mutates fun skipBits(self: Slice, l: Int) { SDSKIPFIRST }
+
+asm extends fun sliceLast(self: Slice, len: Int): Slice { SDCUTLAST }
+
+asm fun sliceDepth(slice: Slice): Int { SDEPTH }
```
This should probably be an extension function

```suggestion
asm extends fun depth(self: Slice): Int { SDEPTH }
```

novusnota commented on stdlib/std/cells.tact:
```diff
@@ -48,6 +48,8 @@ asm extends fun refs(self: Builder): Int { BREFS }
 
 asm extends fun bits(self: Builder): Int { BBITS }
 
+asm fun builderDepth(builder: Builder): Int { BDEPTH }
```
This should probably be an extension function

```suggestion
asm extends fun depth(self: Builder): Int { BDEPTH }
```

novusnota commented on stdlib/std/cells.tact:
```diff
@@ -147,4 +157,19 @@ inline fun emptyCell(): Cell {
\ No newline at end of file
+}
+
+asm fun stringHash(data: Slice): Int { SHA256U }
```
This one doesn't make much sense, since in Tact `String` and `Slice` are considered different types, while the `Slice` already has the `.hash()` method which uses the same TVM instruction. And for both that method and this function, the user would have to do explicit conversion `theirString.asSlice()`, so this function `stringHash()` is probably redundant and should be removed :)

Gusarich replied:
yes, theres `.hash` method for `String` type as well. Removed the function.

novusnota reviewed (commented):
Sorry for reviewing the draft, I just wanted to keep those notes around :)

Gusarich committed:
feat: apply suggestions from code review

Co-authored-by: Novus Nota <68142933+novusnota@users.noreply.github.com>

Gusarich committed:
chore: fix varuint comment in stdlib

Gusarich committed:
feat: remove `stringHash` due to existence of `.hash` method

Gusarich committed:
feat: update `stdlib.ts`

Gusarich committed:
feat: add `MULRSHIFT` functions

Gusarich committed:
feat: update `stdlib.ts`

Gusarich committed:
feat: add `sqrt` function

Gusarich commented:
@novusnota @anton-trunov Implemented all functions. I'd like to hear some intermediate feedback on them before I proceed with writing tests.

I made sure to include all missing functions from both stdlib.fc and mathlib.fc with some exceptions (see PR description).

anton-trunov commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - The `replace` and `replaceGet` methods for the `Map` type: PR [#941](https://github.com/tact-lang/tact/pull/941)
 - Utility for logging errors in code that was supposed to be unreachable: PR [#991](https://github.com/tact-lang/tact/pull/991)
 - Docs: `preloadRef` method for the `Slice` type: PR [#1044](https://github.com/tact-lang/tact/pull/1044)
+- New functions in stdlib from `stdlib.fc` and `math.fc`: PR [#986](https://github.com/tact-lang/tact/pull/986)
```
Let's list all the new functions here -- this is super-useful for editor plugins

Gusarich replied:
done

anton-trunov commented on stdlib/std/cells.tact:
```diff
@@ -191,3 +201,16 @@ inline fun emptyCell(): Cell {
+    refs: Int;
+}
+
+asm fun computeDataSize(cell: Cell, maxCells: Int): DataSize { CDATASIZE TRIPLE }
```
Would `TRIPLE` work here? Structures are returned as tensors, right?

Gusarich replied:
`TRIPLE` makes three elements into a tuple of length 3. I think it should work

anton-trunov replied:
yes, but our structs are tensors, not tuples, right?

Gusarich replied:
structs are implemented as tuples in Tact (that's why we had problems with long ones in the first place, remember?)

anton-trunov replied:
ah, I must have mistaken it with the unbundled approach in the storage

anton-trunov reviewed (commented):
afair, `mathlib.fc` has lots of functions for fixed-arithmetic as well, I think we should add those too (perhaps as a separate library, like `fixed_point_math.tact`).

Also, I left a couple comments in the source code

Gusarich committed:
feat: `sliceLast` -> `lastBits`, and make `computeDataSize` an extend function for both cell and slice

Gusarich committed:
chore: add all function names to changelog

Gusarich committed:
feat: `addressNone`

novusnota commented on stdlib/std/math.tact:
```diff
@@ -59,4 +59,18 @@ native log(num: Int, base: Int): Int;
+asm fun sign(x: Int): Int { SGN }
+
+// floor(x*y/2^z)
+asm fun mulrshift(x: Int, y: Int, z: Int): Int { MULRSHIFT }
```
```suggestion
asm fun mulShiftRight(x: Int, y: Int, z: Int): Int { MULRSHIFT }
```

Not sure if that'll make it less recognizable to users that know about `MULRSHIFT` instruction, but it'll definitely make the function more [accessible to others](https://docs.tact-lang.org/book/operators/#binary-bitwise-shift-right) :)

Similar functions can be renamed in similar fashion

Gusarich replied:
done

anton-trunov commented on src/generator/writers/writeStdlib.ts:
```diff
@@ -1133,6 +1133,32 @@ export function writeStdlib(ctx: WriterContext): void {
+        ctx.asm("", "UBITSIZE");
+    });
+
+    ctx.fun(`__tact_sqrt`, () => {
```
how about we implement this in Tact?

Gusarich replied:
and what about other stdlib functions?

anton-trunov replied:
if a function can be implemented efficiently in Tact, we should implement it in Tact or open an issue and fix it

Gusarich replied:
done

anton-trunov commented on src/generator/writers/writeStdlib.ts:
```diff
@@ -1133,6 +1133,32 @@ export function writeStdlib(ctx: WriterContext): void {
+                int x = (s == 1 ? (num - 1) / 2 + 1 : 1 << ((s + 1) / 2));
+
+                do {
+                    int q = (muldivc(num, 1, x) - x) / 2;
```
looks like `muldivc(num, 1, x)` can be replaced with `divc`

Gusarich replied:
done

Gusarich committed:
feat: rename mulrshift functions

Gusarich committed:
feat: rewrite sqrt on tact

Gusarich committed:
chore: update serialization snapshots

Gusarich committed:
chore: update cspell ignorelist

novusnota commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - The `replace` and `replaceGet` methods for the `Map` type: PR [#941](https://github.com/tact-lang/tact/pull/941)
 - Utility for logging errors in code that was supposed to be unreachable: PR [#991](https://github.com/tact-lang/tact/pull/991)
 - Docs: `preloadRef` method for the `Slice` type: PR [#1044](https://github.com/tact-lang/tact/pull/1044)
+- New functions in stdlib from `stdlib.fc` and `math.fc`: `Builder.depth`, `Slice.skipLastBits`, `Slice.firstBits`, `Slice.lastBits`, `Slice.depth`, `Cell.computeDataSize`, `Slice.computeDataSize`, `Cell.depth`, `curLt`, `blockLt`, `setGasLimit`, `getSeed`, `setSeed`, `myCode`, `sign`, `mulrshift`, `mulrshift_round`, `mulrshift_ceil`, `sqrt`, `addressNone`: PR [#986](https://github.com/tact-lang/ ...
```
```suggestion
- New functions in stdlib from `stdlib.fc` and `math.fc`: `Builder.depth`, `Slice.skipLastBits`, `Slice.firstBits`, `Slice.lastBits`, `Slice.depth`, `Cell.computeDataSize`, `Slice.computeDataSize`, `Cell.depth`, `curLt`, `blockLt`, `setGasLimit`, `getSeed`, `setSeed`, `myCode`, `sign`, `divc`, `muldivc`, `mulShiftRight`, `mulShiftRightRound`, `mulShiftRightCeil`, `sqrt`, `addressNone`: PR [#986](https://github.com/tact-lang/tact/pull/986)
```

Neat!

Gusarich committed:
chore: update CHANGELOG.md

Co-authored-by: Novus Nota <68142933+novusnota@users.noreply.github.com>

Gusarich committed:
fix: `addressNone` and `computeDataSize` types

Gusarich committed:
chore: add comments with specs for new math functions

Gusarich committed:
feat: cover new functions with tests

Gusarich commented:
@anton-trunov @novusnota covered all new functions with tests!

Found and fixed a few little mistakes in types (for `addressNone` and `computeDataSize` functions). Also added comments with formulas for new math functions

anton-trunov commented:
@Gusarich Let's
- resolve the merge conflicts
- move the notes from "current state" to their own separate issues

anton-trunov reviewed (changes requested):
We should document the new functions in https://docs.tact-lang.org/ref/core-math and other relevant doc pages

novusnota commented:
> We should document the new functions in https://docs.tact-lang.org/ref/core-math and other relevant doc pages

Will do

Gusarich committed:
feat: update stdlib.ts

Gusarich commented:
@anton-trunov rebased

Shvandre commented on src/test/e2e-emulated/contracts/stdlib.tact:
```diff
@@ -58,4 +58,70 @@ contract StdlibTest {
+        return sc.depth();
+    }
+
+    get fun addressNone(): Address? {
```
Why does it return "Maybe Address" and not "Address" ?

Gusarich replied:
Because the way Tact currently work with addresses is that "null" as an address is actually addressNone

Shvandre replied:
Well, I checked the sources and you are 100% right, but maybe we should change it?
Or change the tlb schemes generation, because 
```tact
struct Test {
    my_addr: Address?;
}

```

results
```
## Test
TLB: `_ my_addr:Maybe address = Test`
Signature: `Test{my_addr:Maybe address}`
```
In .md file (and maybe in some other places)

Gusarich replied:
it's an old problem in Tact, which was blocking multiple issues in the past. I remember some old comments in address-related issues about that.

My opinion is that we should rework addresses, because currently the way they work is not intuitive and doesn't fit into how other types work.
Basically, `addr_none` should be a valid part of `Address` type and should be checked by devs against `addressNone()` (and not `null`). And `Address?` should mean `Maybe Address`, which also could contain `addr_none`.
@anton-trunov what do you think?

anton-trunov replied:
> addr_none should be a valid part of Address type and should be checked by devs against addressNone() (and not null).

yeah, this is how it should work, indeed

> And Address? should mean Maybe Address

also agreed

> which also could contain addr_none.

this one I'm not sure about, I'd say addr_none in the optional version should be `000`, what do you think?

Gusarich replied:
> this one I'm not sure about, I'd say addr_none in the optional version should be `000`, what do you think?

well yes, that's what I meant. I was talking about the possibility of the case when value of `Address?` is equal to `addressNone()`.

anton-trunov replied:
@Gusarich Maybe you could take care of this change first (allowing `addr_none` as an inhabitant of the `Address` type) and then we merge this PR with the corresponding changes

anton-trunov replied:
We could also probably add a config option to turn off address verification along the way

Shvandre commented on src/test/e2e-emulated/contracts/stdlib.tact:
```diff
@@ -58,4 +58,70 @@ contract StdlibTest {
+        let x = 0;
+        repeat (100) {
+            x += 1;
+        }
```
Maybe I just do not understand something, but why do we need this loop here?

Shvandre replied:
Ah, okay, just for testing purposes.

Shvandre commented on stdlib/std/cells.tact:
```diff
@@ -210,3 +220,18 @@ inline fun emptyCell(): Cell {
     return emptyCell().asSlice();
 }
+
+asm fun addressNone(): Address? { b{00} PUSHSLICE }
```
Same comment as above "Why Maybe Address instead of Address"

Shvandre commented on stdlib/std/math.tact:
```diff
@@ -49,14 +49,52 @@ asm fun abs(x: Int): Int { ABS }
+    do {
+        q = (divc(num, x) - x) / 2;
+        x += q;
+    } until (q == 0);
```
What is the exact method used here? Googled and didn't find anything similar to it

Gusarich replied:
https://github.com/tact-lang/tact/issues/770#issuecomment-2481863626

Shvandre replied:
Perfect

novusnota commented:
> We should document the new functions in https://docs.tact-lang.org/ref/core-math and other relevant doc pages

Halfway there, will be finishing this soon.

UPD: 2/3rds are done, got to complete the cell related functions
UPD: done!

novusnota committed:
chore: typos and extra linkage

novusnota committed:
feat(docs): all math functions and all new non-cell related functions documented, except for `addressNone`

Gusarich commented:
> move the notes from "current state" to their own separate issues

@anton-trunov have a look at the updated notes. there are few of them that I don't think we need to cover in Tact

anton-trunov commented:
Looks good! Thanks for adding those

novusnota committed:
feat(docs): new cell-related functions plus minor corrections here and there

novusnota commented:
@anton-trunov @Gusarich 

Added documentation for all of the new functions from this PR, except for the `addressNone()`, since its implementation is still in the works. Didn't add doc comments yet — once we agree on the descriptions in the documentation, then I'll quickly add the doc comments mirroring them.

Shvandre commented on docs/src/content/docs/ref/core-math.mdx:
```diff
@@ -59,13 +61,169 @@ abs(-42);       // 42
+muldivc(-4, 1, 2); // -2
+muldivc(-3, 1, 2); // -1
+muldivc(-3, 0, 2); // 0
+muldivc(-3, 0, ); // ERROR! Exit code 4: Integer overflow
```
Why there is no 3rd parameter?

novusnota replied:
A typo, thanks for spotting it!

novusnota committed:
chore: typo

Shvandre commented on docs/src/content/docs/ref/core-math.mdx:
````diff
@@ -59,13 +61,169 @@ abs(-42);       // 42
+fun sign(x: Int): Int;
+```
+
+Computes and returns the sign of the [`Int{:tact}`][int] value `x`. Produces $1$ is the `x` is negative, $-1$ if the `x` is positive, and $0$ if the `x` is $0$.
````
I think it produces 1 if x is positive, -1 if x is negative ....

novusnota replied:
Yup, but examples were correct :)

novusnota committed:
chore: typo

anton-trunov commented on src/test/e2e-emulated/contracts/math.tact:
do we want to also add `muldiv`?

novusnota replied:
Probably not, since it seems to be the same as just doing `(x * y) / z` with regular operators

anton-trunov commented:
Let's bring this up to date, resolve the review comments and merge

verytactical commented:
Maybe we should split it into smaller PRs? Given the amount of discussion it created, it would probably be faster to merged if we didn't block merging the code with discussions on an unrelated part of stdlib.

An uneducated guess would be 3-4 PRs:

![image](https://github.com/user-attachments/assets/1c81d6f5-f8fd-44a9-871b-50fc2cf46cc6)

novusnota commented:
> Maybe we should split it into smaller PRs? Given the amount of discussion it created, it would probably be faster to merged if we didn't block merging the code with discussions on an unrelated part of stdlib.

Well, the only unresolved thing here besides merge conflics is the `addressNone()` function — other functions were reviewed and even have public docs written for them right in this PR already.

Therefore, I suggest @Gusarich you (or I) fix merge conflics, add the doc comments to stdlib based on the documentation I've added prior, and then the `addressNone()` will wait for completion as per #1029

@verytactical @anton-trunov

Gusarich committed:
Merge branch 'main' into stdlib-extension-with-mathlib

Gusarich committed:
feat(cspell): add `MULDIVC`

Gusarich commented:
@novusnota @anton-trunov @Shvandre have a look please!

Shvandre commented on docs/src/content/docs/ref/core-cells.mdx:
````diff
@@ -84,6 +84,64 @@ let c: Cell = emptyCell();
+<Badge text="Available since Tact 1.6" variant="tip" size="medium"/><p/>
+
+```tact
+extends fun depth(self: Cell): Int;
````
Maybe the argument should be optional? `extends fun depth(self: Cell?)`. I haven't tested, but TVM.pdf (and TVM spec also) states

> D765 — CDEPTH (c – x), returns the depth of Cell c. If c has no
references, then x = 0; otherwise x is one plus the maximum of depths
of cells referred to from c. If c is a Null instead of a Cell, returns zero.

novusnota replied:
Good point, but calling `null.depth()` in order to simply obtain 0 doesn't make much sense. And if the user has something like `let x: Cell? = ...` at any point, I doubt calling `x.depth()` will be an efficient way to determine if `x` is null or not, since the `!!` would still be needed for later uses that require `Cell` and not `Cell?` type.

But I might be totally wrong :)

Shvandre replied:
I'd say it makes sense if your Cell? is fro, incoming message, and your logic is somehow based on that. I haven't seen those examples, but I would add that as there is no overhead, and internal TVM instruction provides support for it.

Gusarich replied:
changed

Shvandre commented on docs/src/content/docs/ref/core-cells.mdx:
````diff
@@ -84,6 +84,64 @@ let c: Cell = emptyCell();
+
+```tact
+let c: Cell = beginCell().storeInt(42, 7).endCell();
+let depth: Int = c.depth;
````
Are round braces missing here? Or it is more like a `field`, not `method`?

novusnota replied:
Yes, they are missing, good catch!

Shvandre commented on docs/src/content/docs/ref/core-cells.mdx:
````diff
@@ -84,6 +84,64 @@ let c: Cell = emptyCell();
+<Badge text="Available since Tact 1.6" variant="tip" size="medium"/><p/>
+
+```tact
+extends fun computeDataSize(self: Cell, maxCells: Int): DataSize;
````
Same comment here. What are the benefits of narrowing the application of a function? `CDATAZISE` works fine with `null`, and returns $x=y=z=0$

Gusarich replied:
@Shvandre changed

Shvandre commented on docs/src/content/docs/ref/core-cells.mdx:
````diff
@@ -367,6 +425,25 @@ let b: Builder = beginCell();
+
+```tact
+let b: Builder = beginCell().storeInt(42, 7);
+let depth: Int = b.depth;
````
same comment about b.depth

Shvandre commented on docs/src/content/docs/ref/core-common.mdx:
```diff
@@ -125,6 +127,10 @@ Creates a new [`Address{:tact}`][p] based on the [`chain` id](https://ton-blockc
 
 This function tries to resolve constant values in [compile-time](/ref/core-comptime) whenever possible.
 
+Attempts to specify an invalid `chain` ID (not $-1$ or $0$) throw an exception with [exit code 136](/book/exit-codes#136): `Invalid address`.
```
Is this behavior remained the same after address-verification removal?

novusnota replied:
Nope, although the const-eval throws an error if it can detect an invalid `chain` ID, i.e. not -1 or 0

novusnota replied:
Edited the texts

Shvandre commented on docs/src/content/docs/ref/core-common.mdx:
```diff
@@ -125,6 +127,10 @@ Creates a new [`Address{:tact}`][p] based on the [`chain` id](https://ton-blockc
 
+Attempts to specify an invalid `chain` ID (not $-1$ or $0$) throw an exception with [exit code 136](/book/exit-codes#136): `Invalid address`.
+
+Attempts to specify an `chain` ID pointing to the masterchain ($-1$) without [masterchain support](/book/masterchain) enabled throw an exception with [exit code 137](/book/exit-codes#137): `Masterchain support is not enabled for this contract`.
```
Same question, but regarding masterchain flag removal

novusnota replied:
Such runtime errors don't occur anymore, although the const-eval still throws a compilation error if it can detect an invalid `chain` ID, i.e. not -1 or 0

novusnota commented:
@Gusarich is `addressNone()` done? That is, may I document it in its current state, returning `Address?`?

If not, we can remove it from the changelog, docs and sources for now. And then make a follow-up PR that'll re-introduce `addressNone()` with changes to the `Address` type itself, as suggested earlier :)

novusnota committed:
docs(stdlib): add doc comments

novusnota commented:
I've applied most of the suggestions from the code review and also added doc comments to everything, but `addressNone()`.

Gusarich committed:
feat: make self type optional for `Cell.depth` and `Cell.computeDataSize`

Gusarich committed:
Merge branch 'main' into stdlib-extension-with-mathlib

Gusarich committed:
feat: remove `myCode` definition from jetton benchmark

novusnota committed:
chore: small corrections

novusnota committed:
chore: temporarily remove `addressNone()`

novusnota commented:
There's an issue for `addressNone()` already: https://github.com/tact-lang/tact/issues/1029. Thus, I've marked this PR as "Towards" it, but not "Closing" it, and removed the `addressNone()` from all places.

Now we can finally merge this PR :tada: 

@Gusarich @anton-trunov @Shvandre

novusnota committed:
Merge branch 'main' into stdlib-extension-with-mathlib

Commit:
Merge branch 'main' into stdlib-extension-with-mathlib

anton-trunov commented:
merge conflicts

novusnota committed:
Merge branch 'main' into stdlib-extension-with-mathlib

novusnota committed:
chore: stdlib.ts

novusnota committed:
fix: typo from merging

anton-trunov commented:
@novusnota merge conflicts still

novusnota commented:
New ones it seems — I've resolved previous ones last time :)
One moment

novusnota committed:
Merge remote-tracking branch 'upstream/main' into stdlib-extension-with-mathlib

novusnota committed:
chore: stdlib.ts

novusnota committed:
Merge branch 'main' into stdlib-extension-with-mathlib

novusnota committed:
feat: marked some functions as gas-expensive

towards https://github.com/tact-lang/tact/issues/1246

novusnota committed:
Merge branch 'main' into stdlib-extension-with-mathlib

novusnota committed:
chore: add a tooltip on hover for gas-expensive marks

novusnota committed:
Merge branch 'main' into stdlib-extension-with-mathlib

novusnota commented:
macOS CIs exceed some time limits even after restarting those in isolation. Other than that, this PR seems to be done, I've even added the couple of "gas-expensive" badges :)

> We should document the new functions in https://docs.tact-lang.org/ref/core-math and other relevant doc pages

:heavy_check_mark:

anton-trunov commented on docs/src/content/docs/ref/core-math.mdx:
````diff
@@ -59,13 +61,169 @@ abs(-42);       // 42
+fun sqrt(num: Int): Int;
+```
+
+Computes the [square root](https://en.wikipedia.org/wiki/Square_root) of the [`Int{:tact}`][int] value `num`. Returns the result rounded to the nearest integer. If there are two equally close integers, rounding is done toward the even one.
````
maybe https://en.wikipedia.org/wiki/Integer_square_root is a better link here?

novusnota replied:
Hmm, our current algo rounds upwards to the nearest even integer, but the integer square root is supposed to floor and we don't do that per se.

For example, our `sqrt(3)` gives 2, but according to integer square root the result should've been 1. See `python3 -c "import math; print(math.isqrt(3))"`

novusnota replied:
https://github.com/tact-lang/tact/issues/1710

anton-trunov commented on docs/src/content/docs/ref/core-math.mdx:
````diff
@@ -59,13 +61,169 @@ abs(-42);       // 42
+fun mulShiftRight(x: Int, y: Int, z: Int): Int;
+```
+
+Computes and returns the [rounded down][round-down] result of `(x * y) / z{:tact}`. It is a more gas-efficient equivalent of doing the [bitwise shift right](/book/operators#binary-bitwise-shift-right) on the result of multiplication of [`Int{:tact}`][int] `x` times [`Int{:tact}`][int] `y`, where [`Int{:tact}`][int] `z` is the right operand of the shift.
````
```suggestion
Computes and returns the [rounded down][round-down] result of `(x * y) / 2^z{:tact}`. It is a more gas-efficient equivalent of doing the [bitwise shift right](/book/operators#binary-bitwise-shift-right) on the result of multiplication of [`Int{:tact}`][int] `x` times [`Int{:tact}`][int] `y`, where [`Int{:tact}`][int] `z` is the right operand of the shift.
```

novusnota replied:
Applied the fix, and in the code comments too

anton-trunov commented on docs/src/content/docs/ref/core-math.mdx:
```diff
@@ -59,13 +61,169 @@ abs(-42);       // 42
+
+If the value in calculation goes beyond the range from $-2^{256}$ to $2^{256} - 1$ inclusive, an exception with [exit code 4](/book/exit-codes#4) is thrown: `Integer overflow`.
+
+Attempts to specify a negative value of `z` throw an exception with [exit code 5](/book/exit-codes#5): `Integer out of expected range`.
```
Let's check if there are other restrictions on `z` like the ones for the shift operations

novusnota replied:
Yup, not just negative, but also values of 257 and onward aren't allowed.

The allowed range is from 0 to 256 inclusive

novusnota replied:
Added that info to docs and code comments

anton-trunov reviewed (changes requested):
Almost done, just one final push

novusnota committed:
chore: address comments from the code review

novusnota commented:
Let's finally merge this :)

i582 commented on src/stdlib/stdlib/std/internal/contract.tact:
```diff
@@ -240,3 +240,106 @@ asm fun parseStdAddress(slice: Slice): StdAddress { REWRITESTDADDR }
+/// * https://docs.tact-lang.org/ref/core-advanced#nativerandomize
+/// * https://docs.tact-lang.org/ref/core-advanced#nativerandomizelt
+///
+asm fun curLt(): Int { LTIME }
```
Let's call it `currentLogicalTime`

novusnota replied:
Hmm, let's not, since we have the `nativeRandomizeLt()` already. And in general people are more or less aware of the meaning of "lt" or at least accustomed to seeing it. 

Besides, the function `now()` is not called `currentTime()`, but just `now()` :)

i582 commented on src/stdlib/stdlib/std/internal/contract.tact:
```diff
@@ -240,3 +240,106 @@ asm fun parseStdAddress(slice: Slice): StdAddress { REWRITESTDADDR }
+///
+/// See: https://docs.tact-lang.org/ref/core-advanced#blocklt
+///
+asm fun blockLt(): Int { BLOCKLT }
```
`blockLogicalTime`

novusnota replied:
Same reasoning as here: https://github.com/tact-lang/tact/pull/986#discussion_r1942642896

i582 commented on src/stdlib/stdlib/std/internal/math.tact:
```diff
@@ -49,14 +49,174 @@ asm fun abs(x: Int): Int { ABS }
 
 asm fun now(): Int { NOW }
 
+// floor(log2(num))
```
Please add more information about what you are describing here

i582 commented on src/stdlib/stdlib/std/internal/math.tact:
```diff
@@ -6,8 +6,8 @@
 
-/// Equivalent to `nativeRandomize(cur_lt())`,
-/// where `cur_lt()` is the logical time of the current transaction
+// Equivalent to `nativeRandomize(cur_lt())`,
```
We (and you too, actually) use `///` for documentation comments :D
Also we don't have `cur_lt()` function in stdlib

novusnota replied:
Yeah, those comments were not touched at all by intent of this PR and I've only explicitly made them `//` since those comments didn't follow the doc comment guidelines :)

But I'll update those too, sure

i582 commented on src/test/e2e-emulated/math.spec.ts:
```diff
@@ -507,6 +507,64 @@ describe("math", () => {
+
+        // sqrt: round(sqrt(num))
+        for (let num = 0n; num <= 100n; num++) {
+            // console.log(`sqrt(${num}) = ${await contract.getSqrt(num)}`);
```
Please, remove commented code

i582 commented on src/stdlib/stdlib/std/internal/cells.tact:
```diff
@@ -382,3 +474,95 @@ inline fun emptyCell(): Cell {
+/// The results are packed into a `DataSize` Struct consisting of:
+/// * `cells: Int`, the total number of nested cells, including the starting one
+/// * `bits: Int`, the total number of bits in all nested cells, including the starting one
+/// * `refs: Int`, the total number of refs in all nested cells, including the starting one
```
These comments should be in the structure itself (for hover documentation)

novusnota replied:
Good point!

UPD: Wait, but wouldn't the structure itself simply have its own declaration present in the hover? That said, I'll add the `///` comments for each of the fields, yeah.

novusnota commented:
Uno momento

novusnota committed:
docs: more doc comments addressing code review

Towards https://github.com/tact-lang/tact/issues/573

novusnota committed:
Merge branch 'main' into stdlib-extension-with-mathlib

novusnota commented:
@i582 Resolved your suggestions (hopefully), please take a look

i582 reviewed (approved):
LGTM

novusnota committed:
Merge branch 'main' into stdlib-extension-with-mathlib

novusnota commented:
Hooray :tada:
